Encoding categorical predictors (#216)
tidymodels/TMwR: 16-encoding-categorical-data.Rmd

library(embed)
library(textrecipes)
library(kableExtra)
tidymodels_prefer()
source("ames_snippets.R")

neighborhood_counts <- count(ames_train, Neighborhood)
```

# (PART\*) Other Topics {-} 

# Encoding categorical data  {#categorical}

In Section \@ref(dummies) we introduced feature engineering approaches to encode or transform qualitative or nominal data into a representation better suited for most model algorithms. We discussed how to transform a categorical variable, such as the `Bldg_Type` in our Ames housing data (with levels `r knitr::combine_words(glue::backtick(levels(ames_train$Bldg_Type)))`), to a set of dummy or indicator variables.

```{r engineering-all-dummies, echo = FALSE, results = 'asis'}
show_rows <- 
  ames_train %>% 
  mutate(.row = row_number()) %>% 
  group_by(Bldg_Type) %>% dplyr::select(Bldg_Type, .row) %>% 
  slice(1) %>% 
  pull(.row)
recipe(~Bldg_Type, data = ames_train) %>% 
  step_mutate(`Raw Data` = Bldg_Type) %>% 
  step_dummy(Bldg_Type, naming = function(var, lvl, ordinal = FALSE, sep = "_") lvl) %>% 
  prep() %>% 
  bake(ames_train) %>% 
  slice(show_rows) %>% 
  arrange(`Raw Data`) %>% 
  kable() %>% 
  kable_styling(full_width = FALSE)
```

> Many model implementations require such a transformation to a numeric representation for categorical data. 

Reviewer: Should we mention that R prefers factors to represent these data and python typically uses integers?

:::rmdnote
Appendix \@ref(pre-proc-table) presents a table of recommended preprocessing techniques for different models; notice how many of the models in the table require a numeric encoding for all predictors.
:::

> However, for many realistic data sets, straightforward dummy variables are not a good fit. This often happens because there are **too many** categories or there are **new** categories at prediction time. In this chapter, we discuss more sophisticated options for encoding categorical predictors that address these issues. These options are available as recipe steps in the [`r pkg(embed)`](https://embed.tidymodels.org/) package.

Reviewer: ```suggestion
However, for some realistic data sets, straightforward dummy variables are not a good fit. This often happens because there are **too many** categories or there are **new** categories at prediction time. In this chapter, we discuss more sophisticated options for encoding categorical predictors that address these issues. These options are available as recipe steps in the [`r pkg(embed)`](https://embed.tidymodels.org/)and [`r pkg(textrecipes)`](https://textrecipes.tidymodels.org/) packages.
```

These models that can handle categorical features natively can _also_ deal with numeric, continuous features, making the transformation or encoding of such variables optional. Does this help? See Section 5.7 of @fes for a detailed exploration of how such models perform on benchmark data sets with untransformed factor variables compared with transformed dummy variables for those same features. In short, using dummy encodings did not typically result in better model performance but often required more time to train the models. 

:::rmdnote
We advise starting with untransformed categorical variables when a model allows it, and point out that more complex encodings often do not result in better performance for such models.
:::

## Using the outcome for encoding predictors

> There are multiple options for encodings more complex than dummy or indicator variables. One method called _effect_ or _likelihood encodings_ replaces the original categorical variables with a single numeric column that measures the effect of those data. For example, for the neighborhood predictor in the Ames housing data, we can compute the mean or median sale price for each neighborhood and substitute these means for the original data values. 

Reviewer: For this (and a few other topics) I'll update the PR with more references.

> ```{r}

Reviewer: I like this a lot more:

```r
ames_train %>%
  group_by(Neighborhood) %>%
  summarize(mean = mean(Sale_Price),
            std_err = sd(Sale_Price) / sqrt(length(Sale_Price))) %>% 
  ggplot(aes(y = reorder(Neighborhood, mean), x = mean)) + 
  geom_point() +
  geom_errorbar(aes(xmin = mean - 1.64 * std_err, xmax = mean + 1.64 * std_err)) +
  labs(y = NULL, x = "Price (mean, log scale)")
```

![image](https://user-images.githubusercontent.com/5731043/150562858-6c0d05d8-40c8-4be2-b348-30b74d92a8b7.png)

  step_lencode_glm(Neighborhood, outcome = vars(Sale_Price)) %>%
  step_dummy(all_nominal_predictors()) %>% 
  step_interact( ~ Gr_Liv_Area:starts_with("Bldg_Type_") ) %>% 
  step_ns(Latitude, Longitude, deg_free = 20)

ames_glm
```

:::rmdnote
When using a recipe step like `step_lencode_glm()`, specify the variable being encoded first and then the outcome using `vars()`.
:::

As detailed in Section \@ref(recipe-functions), we can `prep()` our recipe to fit or estimate parameters for the preprocessing transformations using training data. We can then `tidy()` this prepared recipe to see the results.

```{r}
glm_estimates <-
  prep(ames_glm) %>%
  tidy(number = 2)

glm_estimates
```

When we use the newly encoded `Neighborhood` numeric variable created via this method, we substitute the original level (such as `"North_Ames"`) with the estimate for `Sale_Price` from the GLM.

Effect encoding methods like this one can also seamlessly handle situations where a novel factor level is encountered in the data. 

```{r}
glm_estimates %>%
  filter(level == "..new")
```

This is the predicted price from the GLM when we don't have any specific neighborhood information.

:::rmdwarn

> Effect encodings can be powerful but should be used with care.

Reviewer: ```suggestion
Effect encodings can be powerful but should be used with care. The effects should be computed from the training set (and not prior to data splitting or modeling). Also, this type of _supervised_ preprocessing should be thoroughly resampled to avoid overfitting. 
```

:::

> When you create an effect encoding for your categorical variable, you are effectively layering a mini-model inside your actual model; this can easily lead to overfitting. To be cautious, this type of encoding should be rigorously resampled (see Chapter \@ref(resampling)). 

Reviewer: I would remove the back end of this paragraph and call it out more vocally in the note above.

> When you create an effect encoding for your categorical variable, you are effectively layering a mini-model inside your actual model; this can easily lead to overfitting. To be cautious, this type of encoding should be rigorously resampled (see Chapter \@ref(resampling)). 

Reviewer: ```suggestion
When you create an effect encoding for your categorical variable, you are effectively layering a mini-model inside your actual model. 
```

## Feature hashing

Traditional dummy variables as described in Section \@ref(dummies) require that all of the possible categories be known to create a full set of numeric features. _Feature hashing_ methods also create dummy variables, but only consider the value of the category to assign it to a predefined pool of dummy variables. Let's look at the `Neighborhood` values in Ames again and use the `rlang::hash()` function to understand more.

```{r}
library(rlang)

ames_hashed <-
  ames_train %>%
  mutate(Hash = map_chr(Neighborhood, hash))

ames_hashed %>%
  select(Neighborhood, Hash)
```

If we input Briardale to this hashing function, we will always get the same output. The neighborhoods in this case are called the "keys", while the outputs are the "hashes".

:::rmdnote
A hashing function takes an input of variable size and maps it to an output of fixed size. Hashing functions are commonly used in cryptography and databases.
:::

> The `rlang::hash()` function generates a 128-bit hash, which means there are `r 2^128` possible hash values. This is great for some applications but doesn't help with feature hashing of **high cardinality** variables (variables with many levels). In feature hashing, the number of possible hashes is a hyperparameter and is set by the model developer through computing the modulo of the integer hashes.

Reviewer: ```suggestion
The `rlang::hash()` function generates a 128-bit hash, which means there are `2^128` possible hash values. This is great for some applications but doesn't help with feature hashing of **high cardinality** variables (variables with many levels). In feature hashing, the number of possible hashes is a hyperparameter and is set by the model developer through computing the modulo of the integer hashes.
```

We can implement feature hashing using a recipe step from the `r pkg(textrecipes)` package.

:::rmdnote
Feature hashing is useful for text data as well as high cardinality categorical data. See Section 6.7 of @Hvitfeldt2021 for a case study demonstration with text predictors.
:::

```{r}
library(textrecipes)
ames_hash <- 
  recipe(Sale_Price ~ Neighborhood + Gr_Liv_Area + Year_Built + Bldg_Type + 
           Latitude + Longitude, data = ames_train) %>%
  step_log(Gr_Liv_Area, base = 10) %>% 
  step_dummy_hash(Neighborhood, signed = FALSE, num_terms = 16L) %>%
  step_dummy(all_nominal_predictors()) %>% 
  step_interact( ~ Gr_Liv_Area:starts_with("Bldg_Type_") ) %>% 
  step_ns(Latitude, Longitude, deg_free = 20)

ames_hash
```

Feature hashing is fast and efficient but has a few downsides. For example, different category values often map to the same hash value. This is called a _collision_ or _aliasing_. How often did this happen with our neighborhoods in Ames?

> ```{r echo=FALSE}

Reviewer: I think that this would be better as a more compact table. 

This isn't great but something like it that is smaller table might work better: 

```r
hash_table <-
  prep(ames_hash) %>% 
  bake(new_data = NULL, starts_with("Neighborhood_")) %>%
  bind_cols(ames_train %>% select(Neighborhood)) %>%
  distinct() %>%
  select(-Neighborhood) %>%
  map_dbl(sum) %>%
  enframe(value = "number of neighborhoods within a feature") %>% 
  count(`number of neighborhoods within a feature`, name = "number of occurrences")
```

hash_table <-
  prep(ames_hash) %>% 
  bake(new_data = NULL, starts_with("Neighborhood_")) %>%
  bind_cols(ames_train %>% select(Neighborhood)) %>%
  distinct() %>%
  select(-Neighborhood) %>%
  map_dbl(sum) %>%
  enframe() 

hash_table %>%
  kable(col.names = c("Hash", "Number of neighborhoods")) %>%
  kable_styling(full_width = FALSE)
```

The number of neighborhoods mapped to each hash value varies between `r min(hash_table$value)` and `r max(hash_table$value)`. All of the hash values greater than one are examples of hash collisions.

> - Feature hashing is not directly interpretable because hash functions cannot be reversed. We can't determine what the input category levels were from the output or if a collision occurred.

Reviewer: ```suggestion
- Feature hashing is not directly interpretable because hash functions cannot be reversed. We can't determine what the input category levels were from the hash value or if a collision occurred.
```

Reviewer: The first sentence is true but the second is not. If you have the original values you can see what categories map to a particular column. You can't look at the hash value itself and know what it maps to, nor can you deconvolute when value was important.

- The number of hash values is a _tuning parameter_ of this preprocessing technique, and you should try several values to determine what is best for your particular modeling approach. A lower number of hash values results in more collisions, but a high number may not be an improvement over your original high cardinality variable.

- Feature hashing can handle new category levels at prediction time, since it does not rely on pre-determined dummy variables.

> - You can reduce hash collisions with a _signed_ hash by using `signed = TRUE`.

Reviewer: ```suggestion
- You can reduce hash collisions with a _signed_ hash by using `signed = TRUE`. This expands the values of one to either +/-1, depending on the sign of the hash. 
```

:::rmdwarn
It is likely that some hash columns will contain all zeros, as we see in this example. We recommend a zero-variance filter via `step_zv()` to filter such columns out.
:::

> ## Entity embeddings

Reviewer: I'm torn about this section. I think that the technique has merit when there are one or more really high cardinality predictors. Look at this example (and the others that we've used) I think that readers might walk away thinking that it isn't effective. 

Maybe, instead of this chapter, give a short overview of tolls (including this) for transforming factors to numbers. It could include some parts about ordinal predictors too.

Author: I moved in the section on ordered factors from the recipes chapter, and made a shorter "overview" last section.
